src/compiler/checker.ts

```diff
@@ -16754,8 +16754,10 @@ namespace ts {
                             return restType;
                         }
                     }
-                    return isNumericLiteralName(name) && getIndexTypeOfContextualType(t, IndexKind.Number) ||
-                        getIndexTypeOfContextualType(t, IndexKind.String);
+                    if (!ignoreIndexSignatureWhenNoProp) {

```

Comment:
Rather than this flag, shouldn't the `getIteratedTypeOrElementType` fallback just be duplicated into this as well?